mkosi/machine.py

```diff
@@ -122,9 +142,71 @@ def run(self, commands: Sequence[str], timeout: int = 900, check: bool = True) -
 
         return process
 
-    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+    def _kill(self, *args: Any, **kwargs: Any) -> None:
         if self._serial:
             self._serial.kill(signal.SIGTERM)
             self.exit_code = self._serial.wait()
             self._serial = None
         self.stack.__exit__(*args, **kwargs)
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        self._kill(*args, **kwargs)
+
+    def reboot(self) -> None:
+        self._kill(None, None, None)
+        self._boot()
+
+
+class MkosiMachineTest(unittest.TestCase):
+    mkosi_tests_no_qemu: bool
+    mkosi_tests_no_nspawn: bool
+    args: Sequence[str]
+    machine: Machine
+    exc_info: Tuple[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]] \
+              = (None, None, None)
+
+    def _callTestMethod(self, method: str) -> None:
+        try:
+            super()._callTestMethod(method) # type: ignore
+        except Exception as e:
+            self.exc_info = exc_info()
+            raise e
+
+    # Mirroring the logic from https://github.com/systemd/systemd/blob/main/src/basic/parse-util.c.
+    @classmethod
+    def _parse_boolean(cls, environment_var: str) -> bool:
+        value = os.getenv(environment_var)
+        if not value:
+            die(f"No value assigned to {environment_var}.")
+
+        if value in ("1", "yes", "y", "true", "t", "on"):
+            return True
+        if value in ("0", "no", "n", "false", "f", "off"):
+            return False
+
+        die(f"No valid value assigned to {environment_var}.")
```

Comment:
Turns out we already have a parse_boolean function in `__init__.py. Let's extend that function with the same values as from the systemd function and reuse it here instead of adding our own function.

General feedback on this function is that reading the environment variable inside the _parse_boolean method means we can't use it for anything else than environment variables. Instead, the right interface is the same as the one from systemd, where `parse_boolean()` takes a string and as a result can be used to parse booleans from environment variable values but also from any other string.